src/EscapeAnalysis.jl

```diff
@@ -207,39 +208,49 @@ function find_escapes(ir::IRCode, nargs::Int)
         for pc in nstmts:-1:1
             stmt = stmts.inst[pc]
 
+            # inlinear already computed effect-freeness of this statement (whether it may throw or not)
+            # and if it may throw, we conservatively escape all the arguments
+            is_effect_free = stmts.flag[pc] & IR_FLAG_EFFECT_FREE ≠ 0
+
             # collect escape information
             if isa(stmt, Expr)
                 head = stmt.head
-                if head === :call
-                    ft = widenconst(argextype(first(stmt.args), ir, sptypes, argtypes))
-                    # TODO implement more builtins, make them more accurate
-                    if ft === Core.IntrinsicFunction # XXX we may break soundness here, e.g. `pointerref`
-                        continue
-                    elseif ft === typeof(isa) || ft === typeof(typeof) || ft === typeof(Core.sizeof) || ft === typeof(Core.:(===))
-                        continue
-                    elseif ft === typeof(ifelse) && length(stmt.args) === 4
-                        f, cond, th, el = stmt.args
-                        info = state.ssavalues[pc]
-                        condt = argextype(cond, ir, sptypes, argtypes)
-                        if isa(condt, Const) && isa(condt.val, Bool)
-                            if condt.val
-                                push!(changes, (th, info))
+                if head === :call # TODO implement more builtins, make them more accurate
+                    if !is_effect_free
+                        # TODO we can have a look at builtins.c and limit the escaped arguments if any of them are not thrown
```

Comment:
I am not very sure about this TODO, do you mean that we can prevent propagating unnecessary escapes if we find the builtin function is not throwing, which means the `IR_FLAG_EFFECT_FREE ` can be set by events other than `throw`?

Reply:
I meant we don't necessarily escape all the arguments here, e.g. `isa(o, t)` doesn't escape `o` even when it `throw`s.

Reply:
I see, thanks for the explanation